dipy/reconst/dki.py

```diff
@@ -2486,6 +2586,172 @@ def split_dki_param(dki_params):
     return evals, evecs, kt
 
 
+def diffusion_kurtosis_odf(dki_params, sphere, alpha=4):
+    """ The diffusion kurtosis estimate of the orientation distribution
+    function (DKI-ODF). This estimates the DKI-ODF in every direction of the
+    given sphere in every voxel of the input data.
+
+    Parameters
+    ----------
+    dki_params : ndarray (x, y, z, 27) or (n, 27)
+        All parameters estimated from the diffusion kurtosis model.
+        Parameters are ordered as follow:
+            1) Three diffusion tensor's eingenvalues
+            2) Three lines of the eigenvector matrix each containing the first,
+               second and third coordinates of the eigenvector
+            3) Fifteen elements of the kurtosis tensor
+    sphere : Sphere class instance.
+        The DKI-ODF is calculated in the vertices of this input.
+    alpha : float, optional
+        Radial weighting power. Default is 4 according to [Jen2014]_ and
+        [Raf2015]_.
+
+    Returns
+    -------
+    kODF : ndarray (x, y, z, g) or (n, g)
+        The DKI-ODF sampled for every voxel of the input data and every g
+        directions of the sphere.
+
+    Notes
+    -----
+    This is based on the DKI-ODF estimate proposed by [Jen2014]_. More
+    information about this method can be found in [Raf2015]_.
+
+    .. [Jen2014] Jensen, J.H., Helpern, J.A., Tabesh, A., (2014). Leading
+        non-Gaussian corrections for diffusion orientation distribution
+        function. NMR Biomed. 27, 202-211. http://dx.doi.org/10.1002/nbm.3053.
+
+    .. [Raf2015] Neto Henriques, R., Correia, M.M., Nunes, R.G., Ferreira,
+        H.A. (2015). Exploring the 3D geometry of the diffusion kurtosis
+        tensor - Impact on the development of robust tractography procedures
+        and novel biomarkers. NeuroImage 111: 85-99.
+        doi:10.1016/j.neuroimage.2015.02.004
+    """
+    # flat data
+    outshape = dki_params.shape[:-1]
+    dki_params = dki_params.reshape((-1, dki_params.shape[-1]))
+
+    # Split data
+    evals, evecs, kt = split_dki_param(dki_params)
+
+    # Initialize AKC matrix
+    V = sphere.vertices
+    kODF = np.zeros((len(kt), len(V)))
+
+    rel_i = _positive_evals(evals[..., 0], evals[..., 1], evals[..., 2])
+    kt = kt[rel_i]
+    evecs = evecs[rel_i]
+    evals = evals[rel_i]
+    kODFi = kODF[rel_i]
+
+    # loop over all relevant voxels
+    for vox in range(len(kt)):
+        # Compute dimensionless tensor U
+        R = evecs[vox]
+        dt = np.dot(np.dot(R, np.diag(evals[vox])), R.T)
+        MD = (dt[0, 0] + dt[1, 1] + dt[2, 2]) / 3
+        U = np.linalg.pinv(dt) * MD
+
+        # loop over all directions
+        sampledODF = np.zeros(len(V))
+        for i in range(len(V)):
+            sampledODF[i] = _dki_odf_core(V[i], kt[vox], U, alpha)
+
+        kODFi[vox] = sampledODF
```

Comment:
it would be good to review this block.  I think this whole block can be managed more efficiently.